### PR TITLE
Support check constraints for UPDATE statement in engine and Delta Lake connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Support check constraints for UPDATE statement in engine and Delta Lake connector

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# General
* Add support for check constraints in `UPDATE` statement. ({issue}`17195`)

# Delta Lake
* Add support for check constraints and column invariants in `UPDATE` statements. ({issue}`17195`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -3255,6 +3250,8 @@ protected Scope visitUpdate(Update update, Optional<Scope> scope)
 
             Scope tableScope = analyzer.analyzeForUpdate(table, scope, UpdateKind.UPDATE);
             update.getWhere().ifPresent(where -> analyzeWhere(update, tableScope, where));
+            analyzeCheckConstraints(table, tableName, tableScope, tableSchema.getTableSchema().getCheckConstraints());
+            analysis.registerTable(table, redirection.getTableHandle(), tableName, session.getIdentity().getUser(), tableScope);
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3254]: Is this related or a pre-existing issue?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -2198,7 +2198,6 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
                     .withRelationType(RelationId.anonymous(), new RelationType(outputFields))
                     .build();
             analyzeFiltersAndMasks(table, targetTableName, new RelationType(outputFields), accessControlScope);
-            analyzeCheckConstraints(table, targetTableName, accessControlScope, tableSchema.getTableSchema().getCheckConstraints());
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:2201]: Why did this get removed from here?

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 3]
```diff
@@ -639,6 +648,11 @@ public PlanNode plan(Update node)
                 // This column is updated...
                 Expression original = orderedColumnValues.get(index);
                 Expression setExpression = coerceIfNecessary(analysis, original, original);
+
+                // Rewrite check constraints based on update assignments
+                // The 'a > b' constraint with 'SET a = 1' assignment is rewritten as '1 > b'
+                constraints = Maps.transformValues(constraints, constraint -> replaceExpression(columnSchema.getName(), constraint, setExpression));
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:654]: This is not safe to do. For instance, if the update assignment is non-deterministic and the constraint references the column multiple times, the transformation is not equivalent.

The constraints should be planned as a filter on top of the subplan that computes the assignments, with the column references adjusted to refer to the outputs of such projection.

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3254]: It's related, not pre-existing.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:2201]: It was unused and it causes duplicated constraints.

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java (hunk 3 above)
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java]: @martint Updated. Please take another look.

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 4]
```diff
@@ -701,6 +721,45 @@ public PlanNode plan(Update node)
...
+    private PlanBuilder addCheckConstraints(PlanBuilder subPlanBuilder, Map<Expression, Expression> constraints, ProjectNode assignmentProject)
+    {
+        if (constraints.isEmpty()) {
+            return subPlanBuilder;
+        }
+
+        for (Map.Entry<Expression, Expression> constraint : constraints.entrySet()) {
+            Expression predicate = new IfExpression(
+                    // When predicate evaluates to UNKNOWN (e.g. NULL > 100), it should not violate the check constraint.
+                    new CoalesceExpression(coerceIfNecessary(analysis, constraint.getValue(), constraint.getValue()), TRUE_LITERAL),
```
Comment (pajaks) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:751]:
```suggestion
                    new CoalesceExpression(coerceIfNecessary(analysis, constraint.getKey(), constraint.getValue()), TRUE_LITERAL),
```

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 5]
```diff
@@ -701,6 +721,45 @@ public PlanNode plan(Update node)
...
+    {
+        return ExpressionTreeRewriter.rewriteWith(
+                new ExpressionRewriter<Void>()
+                {
+                    @Override
+                    public Expression rewriteIdentifier(Identifier node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+                    {
+                        if (columnNameToAssignment.containsKey(node.getValue().toLowerCase(ENGLISH))) {
+                            return columnNameToAssignment.get(node.getValue().toLowerCase(ENGLISH));
+                        }
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:734]: The planner shouldn't be doing any identifier-based resolution. Resolving and matching identifiers involves understanding delimited/non-delimited identifier matching rules and is the job of the analyzer. The analyzer should then record whichever annotations or supporting structures necessary for the planner to do its job without looking at identifier semantics.

Also, we're in the process of decoupling the AST from the IR, so it will no longer be possible to rewrite expressions in this manner as the AST domain and IR domain will be separate hierarchies.

Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java]: Moved to `StatementAnalyzer` and changed to rewrite using `Identifier` instead of String column name. Please take another look.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -4773,6 +4777,24 @@ private void analyzeCheckConstraint(Table table, QualifiedObjectName name, Scope
             analysis.addCheckConstraints(table, expression);
         }
 
+        private static Expression rewriteToAssignment(Map<Identifier, Expression> assignments, Expression sourceExpression)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4780]: This is still not the right approach. There analyzer should not synthesize expressions that didn't come from the original query.

Instead, we need to:
* During analysis, analyze constraints (types, etc) and record them without modification. We may need to add a scope to represent the frame of reference that will be used to evaluate the expressions.
* During planning, right before the insert/update/etc operation, add a projection that evaluates the constraint for each column that's being written. The constraint should be rewritten using the facilities provided by the planner (`PlanBuilder`, `withScope`, etc).

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 7]
```diff
@@ -628,12 +630,15 @@ public PlanNode plan(Update node)
...
         Metadata metadata = plannerContext.getMetadata();
         ImmutableList.Builder<Expression> rowBuilder = ImmutableList.builder();
+        Assignments.Builder assignments = Assignments.builder();
 
         // Add column values to the rowBuilder - - the SET expression value for updated
         // columns, and the existing column value for non-updated columns
         for (int columnIndex = 0; columnIndex < mergeAnalysis.getDataColumnHandles().size(); columnIndex++) {
             ColumnHandle dataColumnHandle = mergeAnalysis.getDataColumnHandles().get(columnIndex);
             ColumnSchema columnSchema = mergeAnalysis.getDataColumnSchemas().get(columnIndex);
+            Integer fieldNumber = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(dataColumnHandle), "Field number for ColumnHandle is null");
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:640]: `int fieldNumber`

Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:640]: Also, remove `requireNonNull`. The null check will be handled by the unboxing of the Integer.

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 8]
```diff
@@ -668,14 +677,40 @@ public PlanNode plan(Update node)
...
+                Expression predicate = new IfExpression(
+                        // When predicate evaluates to UNKNOWN (e.g. NULL > 100), it should not violate the check constraint.
+                        new CoalesceExpression(coerceIfNecessary(analysis, symbol, symbol), TRUE_LITERAL),
+                        TRUE_LITERAL,
+                        new Cast(failFunction(plannerContext.getMetadata(), session, CONSTRAINT_VIOLATION, "Check constraint violation: " + constraint), toSqlType(BOOLEAN)));
+
+                predicates.add(predicate);
+            }
+
+            if (!predicates.isEmpty()) {
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:702]: This check is not needed given the `if (!constraints.isEmpty())` check above.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java [hunk 9]
```diff
@@ -104,6 +104,39 @@ public static Object[][] checkConstraints()
         };
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_73, PROFILE_SPECIFIC_TESTS})
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
```
Comment (martint) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:108]: Why are we introducing a flaky test?

Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:108]: All Databricks product tests need this annotation because they throw 503 error occasionally. We decided to retry the test method instead of adding retry logic to `onDelta` method because DML could succeed in Databricks side even if they return the error.

**Comment by ebyhr** [comment]
@martint Addressed comments.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java (hunk 9 above)
Comment (martint) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:108]: That's a big code smell IMO. The purpose of annotating test as flaky is to avoid spurious test failures until someone goes and improves the test. It should not be used as a mechanism for automatically retrying tests.

If the underlying infrastructure is unreliable, then the test needs to account for that by having a retry mechanism *in the test* itself.

Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:108]: It's pre-existing issue and I don't think starting the discussion in this PR is a good idea. Let us discuss the topic internally.

Comment (martint) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:108]: In that case, I would argue we shouldn't add the test until we can test it properly, or unless we're planning to immediately follow up with a fix. We should avoid accumulating technical debt.

Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java]: Removed `@Flaky` annotatin, `DELTA_LAKE_DATABRICKS` and `DELTA_LAKE_EXCLUDE_73` groups.

**Review by martint — APPROVED** [review]
